### PR TITLE
chore: refactor typesystem function that flattens a userset, and add tests

### PR DESCRIPTION
**Pull request by senojj**
refactor `tupleToUsersetsDefinitions` to `flattenUserset` and add tests.

## Description
adds a comprehensive set of tests for function

## Review Checklist
- [ ] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [x] The correct base branch is being used, if not `main`
- [x] I have added tests to validate that the change in functionality is working as expected

### CODE REVIEW

File: pkg/typesystem/typesystem_test.go [hunk 1]
```diff
@@ -17,6 +17,205 @@ type relationDetails struct {
...
+func TestFlattenUserset(t *testing.T) {
+	tests := map[string]struct {
+		input *openfgav1.Userset
+		o     []*openfgav1.TupleToUserset
+	}{
+		"nil": {
+			input: nil,
+			o:     []*openfgav1.TupleToUserset{},
+		},
+		"nil_userset": {
```
[the remark this answers is not in the document]
Comment (senojj) [reply pkg/typesystem/typesystem_test.go:29]: `nil_userset` indicates that the `Userset` property of the input will be `nil`. some of the following case names appear to be driven by dyslexia. I'll fix.

**Review by adriantam — APPROVED** [review]

File: pkg/typesystem/typesystem.go [hunk 2]
```diff
@@ -1203,25 +1202,33 @@ func (t *TypeSystem) IsTuplesetRelation(objectType, relation string) (bool, erro
...
+	switch x := userset.(type) {
+	case *openfgav1.Userset_TupleToUserset:
+		if x.TupleToUserset != nil {
+			output = append(output, x.TupleToUserset)
 		}
-	}
-	if relationDef.GetIntersection() != nil {
-		for _, child := range relationDef.GetIntersection().GetChild() {
-			tupleToUsersetsDefinitions(child, resp)
+	case *openfgav1.Userset_Union:
```
Comment (miparnisari) [inline pkg/typesystem/typesystem.go:1213]: should we not have a case for `computedUserset` too? 😬

Comment (senojj) [reply pkg/typesystem/typesystem.go:1213]: I was thinking the same thing, but didn't know for sure

Comment (senojj) [reply pkg/typesystem/typesystem.go:1213]: You are right, there are certainly better ways of handling this. I have a hunch that I may be able to reuse some existing code to get this answer more efficiently. Stay tuned.

**Review by miparnisari — APPROVED** [review]

File: pkg/typesystem/typesystem.go (hunk 2 above)
Comment (miparnisari) [reply pkg/typesystem/typesystem.go:1213]: Don't have to do this now, open an issue 👍

Comment (elbuo8) [reply pkg/typesystem/typesystem.go:1213]: I don't see any reason why we can't support `parent` being a computed userset as long as it resolves to a direct relation. Not saying that we should resolve it in this PR but perhaps a nice to have. cc @rhamzeh

**Pull request merged**